File: python/examples/10_jacobi_relative_accuracy.py
# %% [markdown]
# # 10. Relative accuracy: why `syev_jacobi_cta` exists
#
# Every tridiagonalisation-based eigensolver — `syev`, `syev_cta`, LAPACK's
# `dsyev`, NumPy's `eigvalsh` — has an error bound proportional to $\lVert A \rVert$:
#
# $$\lvert \hat\lambda_i - \lambda_i \rvert \lesssim \varepsilon \, \lVert A \rVert$$
#
# For a badly scaled matrix $\lVert A \rVert$ is set by the *largest* eigenvalue,
# so the tiny eigenvalues come out with no correct digits at all.
#
# Two-sided Jacobi is different. For symmetric positive definite input its error
# is governed by the condition number of the **column-equilibrated** matrix
# rather than of $A$ itself (Demmel & Veselic, *SIMAX* 13(4), 1992), so it gets
# the small eigenvalues right too.
#
# That is what `syev_jacobi_cta` provides, for $n \le 32$.

# %%
import numpy as np

import batchlas as bl

from _common import header, preferred_device, report, section


def graded_spd(n: int, batch: int, spread: float, seed: int = 0) -> tuple[np.ndarray, np.ndarray]:
    """Build A = D S D with a huge diagonal grading, plus a reference spectrum.

    S is a well-conditioned SPD matrix (eigenvalues in [1, 2]) and D is diagonal
    with entries spanning ``10**spread``. The result is SPD and badly scaled.

    numpy.linalg.eigvalsh is itself a tridiagonalisation-based solver, so it is
    not automatically a trustworthy reference for tiny eigenvalues. For this
    particular D-S-D construction it does retain relative accuracy, and the fact
    that it agrees with the Jacobi solver to ~1e-13 below -- while the other
    tridiagonal solver does not -- cross-validates both.
    """
    rng = np.random.default_rng(seed)
    matrices = []
    for _ in range(batch):
        q, _ = np.linalg.qr(rng.standard_normal((n, n)))
        core = q @ np.diag(np.linspace(1.0, 2.0, n)) @ q.T
        core = (core + core.T) / 2.0
        scale = np.logspace(0, -spread, n)
        matrices.append(np.diag(scale) @ core @ np.diag(scale))
    a = np.stack(matrices)
    a = (a + a.transpose(0, 2, 1)) / 2.0
    return a, np.stack([np.linalg.eigvalsh(item) for item in a])


def relative_error(computed: np.ndarray, reference: np.ndarray) -> np.ndarray:
    return np.abs(np.sort(computed, axis=-1) - np.sort(reference, axis=-1)) / np.abs(
        np.sort(reference, axis=-1)
    )


header("10. Relative accuracy: syev_jacobi_cta vs the tridiagonal path")

device = preferred_device()
n, batch = 24, 4

# %% [markdown]
# ## On a well-scaled matrix, both are fine
#
# There is no accuracy argument for Jacobi here — this is the baseline.

# %%
section("Both solvers agree on a well-scaled matrix")

rng = np.random.default_rng(1)
plain = rng.standard_normal((batch, n, n))
plain = plain @ plain.transpose(0, 2, 1) + n * np.eye(n)
reference = np.linalg.eigvalsh(plain)

for name in ("syev_cta", "syev_jacobi_cta"):
    values = getattr(bl, name)(plain, compute_vectors=False, device=device)
    report(f"{name:16s} max relative error", float(relative_error(values, reference).max()), tol=1e-12)

# %% [markdown]
# ## A graded matrix
#
# $A = D S D$ where $D$ spans ten orders of magnitude. Note how far apart the
# largest and smallest eigenvalues are.

# %%
section("A graded matrix: A = D S D with D spanning 10 orders of magnitude")

graded, graded_reference = graded_spd(n, batch, spread=10.0, seed=2)

report("condition number", float(np.linalg.cond(graded[0])))
report("largest eigenvalue", float(graded_reference[0].max()))
report("smallest eigenvalue", float(graded_reference[0].min()))

# %% [markdown]
# ## Absolute error hides the problem entirely
#
# Absolute error is bounded by $\varepsilon \lVert A \rVert$, and $\lVert A \rVert$
# is dominated by the large eigenvalues. By this metric both solvers look perfect.

# %%
section("Absolute error: both look fine")

for name in ("syev_cta", "syev_jacobi_cta"):
    values = getattr(bl, name)(graded, compute_vectors=False, device=device)
    absolute = float(np.abs(np.sort(values, axis=-1) - np.sort(graded_reference, axis=-1)).max())
    report(f"{name:16s} max absolute error", absolute)

# %% [markdown]
# ## Relative error tells the real story
#
# This is the number that decides whether the small eigenvalues carry any
# information at all. Watch the *smallest* eigenvalue in particular.

# %%
section("Relative error on the smallest eigenvalues: the real difference")

for name in ("syev_cta", "syev_jacobi_cta"):
    values = getattr(bl, name)(graded, compute_vectors=False, device=device)
    errors = relative_error(values, graded_reference)
    report(f"{name:16s} relative error, smallest eigenvalue", float(errors[:, 0].max()))
    report(f"{name:16s} relative error, largest eigenvalue", float(errors[:, -1].max()))

# %% [markdown]
# ## Sweeping the grading
#
# As the grading grows, the tridiagonal path loses the small eigenvalues
# completely while Jacobi stays at machine precision throughout.

# %%
section("Sweeping the grading")

for spread in (2.0, 4.0, 6.0, 8.0, 10.0):
    graded, graded_reference = graded_spd(n, 2, spread=spread, seed=3)
    line = []
    for name in ("syev_cta", "syev_jacobi_cta"):
        values = getattr(bl, name)(graded, compute_vectors=False, device=device)
        line.append(float(relative_error(values, graded_reference)[:, 0].max()))
    report(f"10^-{spread:<4.0f} grading  syev_cta / jacobi", f"{line[0]:.3e}  /  {line[1]:.3e}")

# %% [markdown]
# ## Tuning the Jacobi sweep
#
# `JacobiOptions` controls the stopping threshold and the sweep cap. Raising
# `tol_multiplier` loosens the off-diagonal threshold, trading accuracy for fewer
# sweeps; on this problem the answer is unchanged across the range.

# %%
section("Tuning the Jacobi sweep")

graded, graded_reference = graded_spd(n, batch, spread=8.0, seed=4)

for multiplier in (1.0, 1e3, 1e6):
    options = bl.JacobiOptions(tol_multiplier=multiplier, max_sweeps=30)
    values = bl.syev_jacobi_cta(graded, compute_vectors=False, options=options, device=device)
    report(
        f"tol_multiplier={multiplier:<8g} relative error (smallest)",
        float(relative_error(values, graded_reference)[:, 0].max()),
    )

# %% [markdown]
# ## Eigenvectors too
#
# The accuracy argument is about eigenvalues, but the solver returns orthonormal
# eigenvectors as well.

# %%
section("Eigenvectors are computed too")

values, vectors = bl.syev_jacobi_cta(graded, device=device)

report("|A V - V diag(w)|", float(np.abs(graded @ vectors - vectors * values[:, None, :]).max()))
report(
    "|V^T V - I|",
    float(np.abs(vectors.transpose(0, 2, 1) @ vectors - np.eye(n)).max()),
    tol=1e-12,
)

File: examples/cpp/10_jacobi_relative_accuracy.cc
// 10. Relative accuracy: why syev_jacobi_cta exists
//
// syev_cta and syev_jacobi_cta solve the same problem and agree to machine
// precision on a well-scaled matrix. On a *graded* one they do not — and
// absolute error hides the difference completely.
//
// The matrix class that matters is A = D S D, with D a diagonal scaling
// spanning many orders of magnitude and S well conditioned. Every entry still
// carries full relative precision, so the tiny eigenvalues are well determined
// by the data — but only an algorithm that never differences numbers of wildly
// different magnitude can recover them. Tridiagonalization does difference
// them; two-sided Jacobi with a relative stopping criterion does not.
// (Demmel & Veselic, SIMAX 13(4), 1992.)
//
// Both solvers are CTA variants, so this example is GPU-only.

#include <cmath>
#include <cstddef>
#include <iomanip>
#include <iostream>

#include <blas/linalg.hh>
#include <util/sycl-vector.hh>

#include "example_utils.hh"

using namespace batchlas;
using namespace examples;

namespace {

constexpr int kN = 16;

// A = D S D. S has a unit diagonal and modest off-diagonal entries; D grades
// the rows and columns from 1 down to 10^-grading.
Matrix<double> graded_matrix(Queue& ctx, double grading, unsigned seed = 42) {
    auto A = Matrix<double>::Random(kN, kN, /*hermitian=*/true, 1, seed);
    ctx.wait();
    for (int i = 0; i < kN; ++i) A(i, i, 0) = 1.0;
    for (int j = 0; j < kN; ++j)
        for (int i = 0; i < kN; ++i)
            if (i != j) A(i, j, 0) *= 0.25;

    for (int j = 0; j < kN; ++j) {
        const double dj = std::pow(10.0, -grading * j / (kN - 1));
        for (int i = 0; i < kN; ++i) {
            const double di = std::pow(10.0, -grading * i / (kN - 1));
            A(i, j, 0) *= di * dj;
        }
    }
    return A;
}

// Print two spectra side by side, smallest first, with the relative gap.
void compare(const UnifiedVector<double>& cta, const UnifiedVector<double>& jacobi, int count = 6) {
    std::cout << "      syev_cta            syev_jacobi_cta      relative difference\n";
    for (int i = 0; i < count; ++i) {
        const double a = cta[i], b = jacobi[i];
        const double rel = (b != 0.0) ? std::abs(a - b) / std::abs(b) : 0.0;
        std::cout << "  " << std::scientific << std::setprecision(6) << std::setw(14) << a << "     " << std::setw(14)
                  << b << "     " << std::setprecision(2) << rel << "\n";
    }
    std::cout << std::defaultfloat;
}

template <Backend B>
struct Example {
    static void run(Queue& ctx) {
        if constexpr (!has_cta_variants<B>) {
            skip("this example", "the CTA variants are not instantiated for this backend");
            return;
        } else {
            if (!supports_cta(ctx)) {
                skip("this example", "needs a GPU with sub-group width 32");
                return;
            }
            well_scaled_section(ctx);
            graded_section(ctx);
            sweep_section(ctx);
            tuning_section(ctx);
            vectors_section(ctx);
            not_automatic_section(ctx);
        }
    }

    static UnifiedVector<double> solve_cta(Queue& ctx, const Matrix<double>& A) {
        auto M = A.clone();
        UnifiedVector<double> w(kN);
        UnifiedVector<std::byte> ws(syev_cta_buffer_size<B>(ctx, M.view(), JobType::NoEigenVectors));
        syev_cta<B>(ctx, M.view(), w.to_span(), JobType::NoEigenVectors, Uplo::Lower, ws.to_span());
        ctx.wait();
        return w;
    }

    static UnifiedVector<double> solve_jacobi(Queue& ctx, const Matrix<double>& A,
                                              JacobiParams<double> params = JacobiParams<double>()) {
        auto M = A.clone();
        UnifiedVector<double> w(kN);
        syev_jacobi_cta<B>(ctx, M.view(), w.to_span(), JobType::NoEigenVectors, Uplo::Lower, Span<std::byte>(),
                           params);
        ctx.wait();
        return w;
    }

    // On a well-scaled matrix, both are fine
    static void well_scaled_section(Queue& ctx) {
        section("On a well-scaled matrix, both agree");

        auto A = graded_matrix(ctx, 0.0);  // D = I
        compare(solve_cta(ctx, A), solve_jacobi(ctx, A), 4);
    }

    // A graded matrix — where they part company
    //
    // The largest eigenvalue sets the norm, so an absolute error of eps*|A| is
    // enormous *relative* to the smallest. Both solvers hit that bound; only
    // one does better.
    static void graded_section(Queue& ctx) {
        section("A graded matrix: D S D with D spanning 1e-7");

        auto A = graded_matrix(ctx, 7.0);
        const auto cta = solve_cta(ctx, A);
        const auto jac = solve_jacobi(ctx, A);

        std::cout << "smallest eigenvalues — the two solvers agree to only a few digits:\n";
        compare(cta, jac, 4);
        std::cout << "largest eigenvalues — here the two agree:\n";
        std::cout << "  syev_cta " << cta[kN - 1] << "   jacobi " << jac[kN - 1] << "\n";
        std::cout << "\nBoth are correct to ~1e-16 times the norm of A, so an absolute-error\n"
                     "check calls them identical. The disagreement is entirely in the digits\n"
                     "an absolute measure cannot see — and it is the tridiagonal path that\n"
                     "loses them, since Jacobi's stopping criterion is relative.\n";
    }

    // Sweeping the grading
    //
    // The gap opens as the scaling widens: identical at grading 0, then orders
    // of magnitude apart.
    static void sweep_section(Queue& ctx) {
        section("Sweeping the grading");

        std::cout << "  grading   smallest from syev_cta   smallest from jacobi\n";
        for (double grading : {0.0, 2.0, 4.0, 6.0, 8.0}) {
            auto A = graded_matrix(ctx, grading);
            const auto cta = solve_cta(ctx, A);
            const auto jac = solve_jacobi(ctx, A);
            std::cout << "  1e-" << std::setw(2) << static_cast<int>(grading) << "     " << std::scientific
                      << std::setprecision(6) << std::setw(14) << cta[0] << "           " << std::setw(14) << jac[0]
                      << std::defaultfloat << "\n";
        }
    }

    // Tuning the Jacobi sweep
    //
    // JacobiParams::tol_multiplier scales the relative threshold a rotation has
    // to exceed to be applied, and max_sweeps caps the cyclic sweeps.
    // Convergence normally takes well under 10 sweeps; cutting it short is what
    // actually costs accuracy.
    static void tuning_section(Queue& ctx) {
        section("Tuning the Jacobi sweep");

        auto A = graded_matrix(ctx, 7.0);

        for (double mult : {1.0, 1e4, 1e8}) {
            JacobiParams<double> params;
            params.tol_multiplier = mult;
            params.max_sweeps = 30;
            const auto w = solve_jacobi(ctx, A, params);
            std::cout << "tol_multiplier = 1e" << static_cast<int>(std::log10(mult)) << ": smallest " << w[0] << "\n";
        }

        for (size_t sweeps : {size_t{1}, size_t{30}}) {
            JacobiParams<double> params;
            params.max_sweeps = sweeps;
            const auto w = solve_jacobi(ctx, A, params);
            std::cout << "max_sweeps = " << sweeps << ": smallest " << w[0] << "\n";
        }
    }

    // Eigenvectors too
    static void vectors_section(Queue& ctx) {
        section("Eigenvectors too");

        auto A = graded_matrix(ctx, 7.0);
        auto M = A.clone();
        UnifiedVector<double> w(kN);
        syev_jacobi_cta<B>(ctx, M.view(), w.to_span(), JobType::EigenVectors, Uplo::Lower);
        ctx.wait();

        // V^T V is the identity when the vectors are orthonormal.
        auto G = Matrix<double>::Zeros(kN, kN, 1);
        gemm<B>(ctx, M, M, G, 1.0, 0.0, Transpose::Trans, Transpose::NoTrans);
        ctx.wait();
        std::cout << "V^T V, top-left corner:\n";
        G.view()[0].print(std::cout, 4, 4);
    }

    // A wide spectrum is not enough
    //
    // Matrix::Random with hermitian = true has a spectrum spread around zero,
    // but it is not of the form D S D. Scale it by a diagonal *after* the fact
    // and the grading is genuine; take an arbitrary matrix with a wide spectrum
    // and it is not. Reach for syev_jacobi_cta because your data is graded, not
    // merely because its eigenvalues span a wide range.
    static void not_automatic_section(Queue& ctx) {
        section("A wide spectrum is not enough");

        // Same eigenvalue range as the graded case, but the smallness is spread
        // over the whole matrix rather than living in a diagonal scaling.
        auto A = Matrix<double>::Random(kN, kN, true, 1, 91);
        ctx.wait();
        for (int j = 0; j < kN; ++j)
            for (int i = 0; i < kN; ++i) A(i, j, 0) *= 1e-7;
        for (int i = 0; i < kN / 2; ++i) A(i, i, 0) += 1.0;

        compare(solve_cta(ctx, A), solve_jacobi(ctx, A), 4);
        std::cout << "Here the two track each other: there is no graded structure for\n"
                     "Jacobi's relative criterion to exploit.\n";
    }
};

}  // namespace

BATCHLAS_EXAMPLE_MAIN("10. Relative accuracy: why syev_jacobi_cta exists")
